File: easy/993/cousins_in_binary_tree.py

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        queue = [root.left, root.right]

        while len(queue) > 0:
            found = 0

            for i in range(len(queue)):
                node = queue.pop(0)
                
                if node:
                    queue.append(node.right)
                    queue.append(node.left)

            for i in range(len(queue)//2):
                n1 = queue.pop(0)
                n2 = queue.pop(0)
                queue.append(n1)
                queue.append(n2)

                if n1 and n2:
                    if n1.val == x or n1.val == y:
                        found += 1 
                    elif n2.val == x or n2.val == y:
                        found += 1
                elif n1:
                    found += 1 if n1.val == x or n1.val == y else 0

                elif n2:
                    found += 1 if n2.val == x or n2.val == y else 0

            if found == 2:
                return True

        return False

File: easy/993/cousins_in_binary_tree.cpp

#include <queue>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    std::queue<TreeNode*> q;
    TreeNode *node, *n1, *n2;
    int found;
    
    bool isCousins(TreeNode* root, int x, int y) {
        q.push(root->left);
        q.push(root->right);
        
        while (!q.empty()) {
            found = 0;

            for (int i = q.size(); i > 0; i--) {
                node = q.front();
                q.pop();
                
                if (node != NULL) {
                    q.push(node->left);
                    q.push(node->right);
                }
            }
            
            for (int i = q.size()/2; i > 0; i--) {
                n1 = q.front();
                q.pop();
                n2 = q.front();
                q.pop();
                q.push(n1);
                q.push(n2);
                
                if (n1 != NULL and n2 != NULL) {
                    if (n1->val == x or n1->val == y) {
                        found += 1;
                    }
                    else if (n2->val == x or n2->val == y) {
                        found += 1;
                    }
                }
                else if (n1 != NULL) {
                    if (n1->val == x or n1->val == y) {
                        found += 1;
                    }
                }
                else if (n2 != NULL) {
                    if (n2->val == x or n2->val == y) {
                        found += 1;
                    }
                }
            }
            
            if (found == 2) {
                return true;
            }
        }

        return false;
    }
};

